### PR TITLE
Cache fulltext search results

### PR DESCRIPTION
**Pull request by OutstandingWork**
Closes # 10051

This PR achives reduction in time taken to show no results by about 0.5-0.8 s

### Technical

- The modification is done in the openlibrary\plugins\openlibrary\code.py
- Added caching for search suggestion responses using `Cache-Control` header
- Set cache duration to 5 minutes (max-age=300)

### Testing

1. Open browser developer tools (F12)
2. Go to Network tab
3. Perform a search query
4. Verify response headers:
   - Should see `Cache-Control: public, max-age=300`
5. Changing the subsection of search
   - Response should be served from browser cache
   - The rendering of loading bar is not there while switching to different subsections (like Authors, Lists etc.)

### Stakeholders

I have worked on this issue with @jimchamp as lead.

### CODE REVIEW

**Review by jimchamp — CHANGES_REQUESTED** [review]
The requirement was to cache the partially rendered HTML that replaces the loading indicator on "no results found" search pages.  You've cached the JSON full-text search results.

[Here](https://github.com/internetarchive/openlibrary/blob/e010b2a13697de70170033902ba2e27a1e1acbe9/openlibrary/plugins/openlibrary/code.py#L1245) is where the data is fetched for the partially rendered HTML.  If those results are error-free, add the `Cache-Control` header.  That is all that is needed.

**Comment by OutstandingWork** [comment]
Should I remove my previous changes and resubmit only the changes needed or should I keep my previous change and the change you suggested?

File: openlibrary/plugins/inside/code.py [hunk 1]
```diff
@@ -41,5 +41,16 @@ def GET(self):
...
+        # Add caching headers only if there's no error
+        if 'error' not in results:
+            # Cache for 5 minutes (300 seconds)
+            web.header('Cache-Control', 'public, max-age=300')
+            web.header('Expires', web.http.expires(300))
+        else:
+            # No caching for error responses
+            web.header('Cache-Control', 'no-store, no-cache, must-revalidate')
+            web.header('Pragma', 'no-cache')
+
```
Comment (jimchamp) [inline openlibrary/plugins/inside/code.py:54]:
```suggestion
```

**Comment by jimchamp** [comment]
Remove the code that you've committed. Only submit what is necessary to close the linked issue.

File: openlibrary/plugins/inside/code.py (hunk 1 above)
Comment (OutstandingWork) [reply openlibrary/plugins/inside/code.py:54]: Screenshot 2024-12-21 010434
Hi @jimchamp , I have done the above changes and is works as expected.

Comment (jimchamp) [reply openlibrary/plugins/inside/code.py:54]: You'll have to push your changes before they can be reviewed and merged.  Remove the `Expires` header before you do so.

Comment (jimchamp) [reply openlibrary/plugins/inside/code.py:54]: Also, remove the `not hits['hits']` part of the conditional.

Comment (OutstandingWork) [reply openlibrary/plugins/inside/code.py:54]: Hi @jimchamp, I have pushed my changes as you told. Please do check once and let me know if the results are as expected.

Comment (jimchamp) [reply openlibrary/plugins/inside/code.py:54]: You have not updated this PR.  Let's try this one last time.  Here is what I am expecting you to do:

1. Update _this_ PR with any new code changes.  No new PR needs to be opened.  Simply push your updates to [the branch that you used to open this PR](https://github.com/OutstandingWork/openlibrary/tree/10051/No_Result_Cache), and this PR will be updated.
2. Immediately following [this line of code](https://github.com/internetarchive/openlibrary/blob/e010b2a13697de70170033902ba2e27a1e1acbe9/openlibrary/plugins/openlibrary/code.py#L1245), add the following two lines of code:
```
if 'error' not in data:
    web.header('Cache-Control', 'public, max-age=300')
```

That is all that is needed.

Comment (OutstandingWork) [reply openlibrary/plugins/inside/code.py:54]: Hi @jimchamp,

Apologies for the confusion earlier—I’m still getting the hang of open-source contributions. I’ve made the requested updates to the PR. If there’s anything I missed or that still needs adjustment, please let me know. Alternatively, feel free to make changes as you see fit. I’m happy to help further if needed.

Thank you for your patience!

Comment (jimchamp) [inline openlibrary/plugins/inside/code.py:54]:
```suggestion
```
Remove unrelated code.

File: openlibrary/plugins/openlibrary/code.py [hunk 2]
```diff
@@ -1243,6 +1243,10 @@ def GET(self):
         elif component == "FulltextSearchSuggestion":
             query = i.get('data', '')
             data = fulltext_search(query)
+            # Add caching headers only if there were no errors in the search results
+            if 'error' not in data:
+                # Cache for 5 minutes (300 seconds)
```
Comment (jimchamp) [inline openlibrary/plugins/openlibrary/code.py:1248]:
```suggestion
```

**Review by jimchamp — APPROVED** [review]
Have added a final commit that removes the unwanted caching on the full-text search `json` endpoint and restores the `return` statement for the `/partials` handler.

@OutstandingWork, thanks for the help, but please familiarize yourself with `git` and GitHub before attempting to contribute to another project.  I recommend reading some primers on both, then perhaps practicing by opening and reviewing PRs for projects in your own repo.

**Comment by OutstandingWork** [comment]
Thanks @jimchamp, will surely familiarize myself with git operations
